File: memcontainers/kernel/rust/src/host_call.rs
//! Kernel-side wrapper over the host-call bridge imports. A guest invokes a
//! host-resident function (the tool broker, a host-backed mount driver) via
//! `mc_sys_host_call`; the host routes the opaque request blob to a registered
//! handler and streams back a result. Exactly like `net::HttpReq`, the raw `i32`
//! handle is a kernel↔host contract that MUST NOT reach the guest — the guest
//! only ever sees an ordinary readable fd. Owns the handle, drives the
//! poll/body/close calls, and closes on `Drop`. Host calls count toward the
//! same in-flight-egress total as the network (a snapshot must not capture a
//! live host handle); see `net::inflight_egress`.

use crate::bridge;
use crate::net::{egress_dec, egress_inc};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HostCallError {
    /// The host refused (no handler / capability denied) — `-1` at start.
    Denied,
    /// A failure after the call started.
    Failed,
}

enum CallPoll {
    Pending,
    Ready,
    Failed,
}

/// An in-flight host call. Owns the host handle; closes it on `Drop`.
struct HostCall {
    handle: i32,
}

impl HostCall {
    fn start(req: &[u8]) -> Result<HostCall, HostCallError> {
        let h = unsafe { bridge::mc_host_call(req.as_ptr(), req.len()) };
        if h < 0 {
            Err(HostCallError::Denied)
        } else {
            egress_inc();
            Ok(HostCall { handle: h })
        }
    }

    fn poll(&mut self) -> CallPoll {
        // The host reports readiness without consuming the body (buf unused).
        let n = unsafe { bridge::mc_host_call_poll(self.handle, core::ptr::null_mut(), 0) };
        if n < 0 {
            CallPoll::Failed
        } else if n == 0 {
            CallPoll::Pending
        } else {
            CallPoll::Ready
        }
    }

    fn read_body(&mut self, out: &mut [u8]) -> Result<usize, HostCallError> {
        let n = unsafe { bridge::mc_host_call_body(self.handle, out.as_mut_ptr(), out.len()) };
        if n < 0 {
            Err(HostCallError::Failed)
        } else {
            Ok((n as usize).min(out.len()))
        }
    }
}

impl Drop for HostCall {
    fn drop(&mut self) {
        unsafe { bridge::mc_host_call_close(self.handle) };
        egress_dec();
    }
}

enum Phase {
    Polling,
    Body,
    Eof,
    Failed,
}

/// Outcome of pulling bytes from a `HostCallSource`.
pub enum HostCallRead {
    /// The host is still computing the result — yield and retry.
    Pending,
    Got(usize),
    Eof,
    Failed,
}

/// A readable host-call result, driven by `mc_sys_read` — poll for readiness
/// (yielding while in flight), then stream the result body, then EOF.
pub struct HostCallSource {
    call: HostCall,
    phase: Phase,
}

impl HostCallSource {
    /// Begin a host call from a request blob; `Denied` if the host refused.
    pub fn start(req: &[u8]) -> Result<HostCallSource, HostCallError> {
        Ok(HostCallSource {
            call: HostCall::start(req)?,
            phase: Phase::Polling,
        })
    }

    pub fn read_into(&mut self, buf: &mut [u8]) -> HostCallRead {
        loop {
            match self.phase {
                Phase::Polling => match self.call.poll() {
                    CallPoll::Pending => return HostCallRead::Pending,
                    CallPoll::Ready => self.phase = Phase::Body,
                    CallPoll::Failed => {
                        self.phase = Phase::Failed;
                        return HostCallRead::Failed;
                    }
                },
                Phase::Body => {
                    return match self.call.read_body(buf) {
                        Ok(0) => {
                            self.phase = Phase::Eof;
                            HostCallRead::Eof
                        }
                        Ok(n) => HostCallRead::Got(n),
                        Err(_) => {
                            self.phase = Phase::Failed;
                            HostCallRead::Failed
                        }
                    };
                }
                Phase::Eof => return HostCallRead::Eof,
                Phase::Failed => return HostCallRead::Failed,
            }
        }
    }

    /// Non-destructive readiness (advance past the poll only).
    pub fn poll_readable(&mut self) -> bool {
        if let Phase::Polling = self.phase {
            match self.call.poll() {
                CallPoll::Pending => return false,
                CallPoll::Ready => self.phase = Phase::Body,
                CallPoll::Failed => self.phase = Phase::Failed,
            }
        }
        true
    }
}

File: memcontainers/kernel/zig/src/egress/host_call.zig
//! src/egress/host_call.zig — opaque host-call operations (§2.8).
//!
//! Owns: the request/poll/complete lifecycle for opaque host calls and their inflight accounting.
//! Invariants: A9, A8 (inflight host calls block snapshots). ASYNC: shallow Pending (§7.4).
//! Oracle (behavior to match): kernel/rust/src/host_call.rs.
//! Not here: the mountfs VFS face (fs/mountfs.zig).
//!
//! Scaffold status: header-only. Fill Phase 6.

const std = @import("std");
const bridge = @import("../bridge.zig");

pub const StartError = error{Denied};

pub const Read = union(enum) {
    pending,
    got: usize,
    eof,
    failed,
};

const Phase = enum {
    polling,
    body,
    eof,
    failed,
};

pub const Engine = struct {
    gpa: std.mem.Allocator,
    inflight_count: i32 = 0,

    pub fn init(gpa: std.mem.Allocator) Engine {
        return .{ .gpa = gpa };
    }

    pub fn inflight(self: *const Engine) i32 {
        return self.inflight_count;
    }

    fn inc(self: *Engine) void {
        self.inflight_count += 1;
    }

    fn dec(self: *Engine) void {
        if (self.inflight_count > 0) self.inflight_count -= 1;
    }

    pub fn start(self: *Engine, req: []const u8) StartError!*Source {
        const src = self.gpa.create(Source) catch @panic("OOM");
        const h = bridge.mc_host_call(req.ptr, req.len);
        if (h < 0) {
            self.gpa.destroy(src);
            return StartError.Denied;
        }
        self.inc();
        src.* = .{
            .engine = self,
            .handle = h,
        };
        return src;
    }
};

pub const Source = struct {
    engine: *Engine,
    handle: i32,
    refs: usize = 1,
    phase: Phase = .polling,

    pub fn retain(self: *Source) *Source {
        self.refs += 1;
        return self;
    }

    pub fn release(self: *Source) void {
        self.refs -|= 1;
        if (self.refs != 0) return;
        const engine = self.engine;
        bridge.mc_host_call_close(self.handle);
        engine.dec();
        engine.gpa.destroy(self);
    }

    fn poll(self: *Source) enum { pending, ready, failed } {
        var unused: [1]u8 = undefined;
        const n = bridge.mc_host_call_poll(self.handle, &unused, 0);
        if (n < 0) return .failed;
        if (n == 0) return .pending;
        return .ready;
    }

    pub fn readInto(self: *Source, out: []u8) Read {
        while (true) {
            switch (self.phase) {
                .polling => switch (self.poll()) {
                    .pending => return .pending,
                    .ready => self.phase = .body,
                    .failed => {
                        self.phase = .failed;
                        return .failed;
                    },
                },
                .body => {
                    const n = bridge.mc_host_call_body(self.handle, out.ptr, out.len);
                    if (n < 0) {
                        self.phase = .failed;
                        return .failed;
                    }
                    const len: usize = @min(@as(usize, @intCast(n)), out.len);
                    if (len == 0) {
                        self.phase = .eof;
                        return .eof;
                    }
                    return .{ .got = len };
                },
                .eof => return .eof,
                .failed => return .failed,
            }
        }
    }

    pub fn pollReadable(self: *Source) bool {
        if (self.phase == .polling) {
            switch (self.poll()) {
                .pending => return false,
                .ready => self.phase = .body,
                .failed => self.phase = .failed,
            }
        }
        return true;
    }
};
